apis/crd/v1alpha1/cloudentityselector_webhook.go

```diff
@@ -218,27 +218,116 @@ func (r *CloudEntitySelector) validateMatchSections() error {
-								" of vpc matchName")
-						}
+			if m.VpcMatch != nil && len(strings.TrimSpace(m.VpcMatch.MatchName)) != 0 {
+				for _, vmMatch := range m.VMMatch {
+					if len(strings.TrimSpace(vmMatch.MatchID)) != 0 ||
+						len(strings.TrimSpace(vmMatch.MatchName)) != 0 {
+						return fmt.Errorf("vpc matchName with either vm matchID" +
+							" or vm matchName is not supported, use vpc matchID instead" +
+							" of vpc matchName")
 					}
 				}
+				if m.Agented {
+					return fmt.Errorf("vpc matchName with agented flag set to true" +
+						"is not supported, use vpc matchID instead")
+				}
+			}
+		}
+	}
+
+	// Block ambiguous match combinations as it can result in unpredictable behavior w.r.t agented configuration.
+	// vpcIDOnlyMatch map holds VPC ID as key if the selector contains only vpcMatch matchID.
+	// vmIDOnlyMatch map holds VM ID as key if the selector contains only vmMatch matchID.
+	// vmNameOnlyMatch map holds VM Name as key if the selector contains only vmMatch matchName.
+	// vmIDWithVpcMatch map holds VM ID and VPC ID as key if the selector contains vpcMatch matchID and vmMatch matchID.
+	// vmNameWithVpcMatch map holds VM Name and VPC ID as key if the selector contains vpcMatch matchID and vmMatch matchName.
+	vpcIDOnlyMatch := make(map[string]struct{})
+	vmIDOnlyMatch := make(map[string]struct{})
+	vmNameOnlyMatch := make(map[string]struct{})
+	vmIDWithVpcMatch := make(map[string]struct{})
+	vmNameWithVpcMatch := make(map[string]struct{})
+	var ret error
+	for _, m := range r.Spec.VMSelector {
+		if m.VpcMatch != nil {
+			ret = r.validateVPCMatch(m, vmIDOnlyMatch, vpcIDOnlyMatch, vmIDWithVpcMatch, vmNameWithVpcMatch)
+		} else {
+			ret = r.validateVMMatch(m, vmIDOnlyMatch, vmNameOnlyMatch)
+		}
+		if ret != nil {
+			return ret
+		}
+	}
+	return nil
+}
+
+// validateVPCMatch function validates VMSelector with VpcMatch section to find if it conflicts with existing VMSelectors.
+// Block same VPC ID configuration in two VMSelectors with only vpcMatch section.
+// Block same VM ID configuration in any two VMSelectors, same is not applicable for VM Name as VM Name need not be unique.
+// Block same combination of VPC ID and VM ID configuration in any two VMSelectors.
+// Block same combination of VPC ID and VM Name configuration in any two VMSelectors.
+func (r *CloudEntitySelector) validateVPCMatch(selector VirtualMachineSelector, vmIDOnlyMatch map[string]struct{},
+	vpcIDOnlyMatch map[string]struct{}, vmIDWithVpcMatch map[string]struct{},
+	vmNameWithVpcMatch map[string]struct{}) error {
+	if selector.VpcMatch.MatchID != "" {
+		if len(selector.VMMatch) == 0 {
+			// Make sure there is no VMSelector with same VPC ID with only vpcMatch section.
+			if _, found := vpcIDOnlyMatch[selector.VpcMatch.MatchID]; found {
+				return fmt.Errorf("same vpcMatch matchID %v configured in two match selectors",
+					selector.VpcMatch.MatchID)
+			}
+			vpcIDOnlyMatch[selector.VpcMatch.MatchID] = struct{}{}
```

Comment:
For the empty struct, can you initialize it to variable and then use it.
we do something like this in aws_ec2.go
exists := struct{}{}

Reply:
done